### PR TITLE
Compile under Visual Studio for ARM64

### PR DESCRIPTION
**Pull request by lemire**
This allows us to build simdjson for ARM64 under Visual Studio. I have not run the tests, but the code builds.

Google Benchmark does not really build for ARM64 under Visual Studio as far as I can tell but we can disable it so it is fine.

There is no possible performance regression, I believe. Performance under Visual Studio is unknown: goal is to compile, period.

### CODE REVIEW

**Comment by lemire** [comment]
I have convinced myself that this does not come with a performance regression on ARM. The `parse` command is this PR while `parsemaster` is from master. 

```
$ ./parse jsonexamples/gsoc-2018.json
number of iterations 200

jsonexamples/gsoc-2018.json
===========================
    51997 blocks -    3327831 bytes - 75841 structurals (  2.3 %)
special blocks with: utf8         0 (  0.0 %) - escape      8253 ( 15.9 %) - 0 structurals     36833 ( 70.8 %) - 1+ structurals     15165 ( 29.2 %) - 8+ structurals      2815 (  5.4 %) - 16+ structurals         0 (  0.0 %)
special block flips: utf8         0 (  0.0 %) - escape     11638 ( 22.4 %) - 0 structurals      7800 ( 15.0 %) - 1+ structurals      7800 ( 15.0 %) - 8+ structurals      3621 (  7.0 %) - 16+ structurals         0 (  0.0 %)

All Stages
|    Speed        : 112.3944 ns per block ( 87.69%) -   1.7562 ns per byte -  77.0597 ns per structural -   0.5694 GB/s
|    Cycles       : 359.7853 per block    ( 99.75%) -   5.6217 per byte    - 246.6755 per structural    -    3.201 GHz est. frequency
|    Instructions : 255.9582 per block    (100.00%) -   3.9994 per byte    - 175.4897 per structural    -    0.711 per cycle
|    Misses       :   38746 branch misses (100.23%) - 113628 cache misses ( 99.99%) - 3268427.00 cache references
|- Stage 1
|    Speed        :  68.6044 ns per block ( 53.52%) -   1.0720 ns per byte -  47.0364 ns per structural -   0.9329 GB/s
|    Cycles       : 219.5893 per block    ( 60.88%) -   3.4311 per byte    - 150.5545 per structural    -    3.201 GHz est. frequency
|    Instructions : 148.0132 per block    ( 57.83%) -   2.3127 per byte    - 101.4806 per structural    -    0.674 per cycle
|    Misses       :    9763 branch misses ( 25.26%) - 52064 cache misses ( 45.82%) - 586832.00 cache references
|- Stage 2
|    Speed        :  43.7118 ns per block ( 34.10%) -   0.6830 ns per byte -  29.9696 ns per structural -   1.4641 GB/s
|    Cycles       : 139.9017 per block    ( 38.79%) -   2.1860 per byte    -  95.9192 per structural    -    3.201 GHz est. frequency
|    Instructions : 107.9157 per block    ( 42.16%) -   1.6862 per byte    -  73.9890 per structural    -    0.771 per cycle
|    Misses       :   28469 branch misses ( 73.65%) - 61542 cache misses ( 54.16%) - 2678510.00 cache references

171.1 documents parsed per second (best)
[dlemire@ampere simdjson]$ ./parsemaster jsonexamples/gsoc-2018.json
number of iterations 200

jsonexamples/gsoc-2018.json
===========================
    51997 blocks -    3327831 bytes - 75841 structurals (  2.3 %)
special blocks with: utf8         0 (  0.0 %) - escape      8253 ( 15.9 %) - 0 structurals     36833 ( 70.8 %) - 1+ structurals     15165 ( 29.2 %) - 8+ structurals      2815 (  5.4 %) - 16+ structurals         0 (  0.0 %)
special block flips: utf8         0 (  0.0 %) - escape     11638 ( 22.4 %) - 0 structurals      7800 ( 15.0 %) - 1+ structurals      7800 ( 15.0 %) - 8+ structurals      3621 (  7.0 %) - 16+ structurals         0 (  0.0 %)

All Stages
|    Speed        : 113.9773 ns per block ( 91.16%) -   1.7809 ns per byte -  78.1450 ns per structural -   0.5615 GB/s
|    Cycles       : 359.1372 per block    ( 99.82%) -   5.6116 per byte    - 246.2311 per structural    -    3.151 GHz est. frequency
|    Instructions : 255.9582 per block    (100.00%) -   3.9994 per byte    - 175.4897 per structural    -    0.713 per cycle
|    Misses       :   38677 branch misses ( 99.90%) - 113560 cache misses ( 99.99%) - 3267635.00 cache references
|- Stage 1
|    Speed        :  69.6805 ns per block ( 55.73%) -   1.0888 ns per byte -  47.7742 ns per structural -   0.9185 GB/s
|    Cycles       : 219.5496 per block    ( 61.02%) -   3.4305 per byte    - 150.5273 per structural    -    3.151 GHz est. frequency
|    Instructions : 148.0132 per block    ( 57.83%) -   2.3127 per byte    - 101.4806 per structural    -    0.674 per cycle
|    Misses       :    9870 branch misses ( 25.49%) - 52063 cache misses ( 45.84%) - 587578.00 cache references
|- Stage 2
|    Speed        :  44.1281 ns per block ( 35.30%) -   0.6895 ns per byte -  30.2550 ns per structural -   1.4503 GB/s
|    Cycles       : 139.0275 per block    ( 38.64%) -   2.1723 per byte    -  95.3199 per structural    -    3.151 GHz est. frequency
|    Instructions : 107.9157 per block    ( 42.16%) -   1.6862 per byte    -  73.9890 per structural    -    0.776 per cycle
|    Misses       :   28704 branch misses ( 74.14%) - 61493 cache misses ( 54.14%) - 2682502.00 cache references

168.7 documents parsed per second (best)
```


Note that since we require a bleeding edge CMake version, and because there is no ARM64 binary distribution, building simdjson with CMake is inconvenient. The workaround is to compile directly...

```
c++ -O3 -o parse benchmark/parse.cpp  -Iinclude -std=c++11 -Ibenchmark/ -Ibenchmark/linux ./src/simdjson.cpp  -Isrc
```

**Comment by lemire** [comment]
I was getting performance warnings. Let us take no chances.

**Comment by lemire** [comment]
@jkeiser  Ok. This is ready for review. The failing perf test on ARM is unexplained BS. I don't think that this PR can possibly be a regression.

So what this PR does is allow you to cross-compile from Windows x64 toward Windows ARM64. It does not test anything, so we cannot claim to support Windows ARM64... but we can at least build simdjson.

This was a painful PR to craft.  I have no idea how you are supposed to initialize a SIMD variable under Visual Studio  for ARM.And the part where `int8x16_t `and `uint8x16_t` are apparently the same type is really odd... The `make_int8x16_t` and `make_uint8x16_t` are some of the ugliest code I have ever written... but I wanted to make sure that it still produced good assembly...  and it is the best I could do.

I was forced to replace

```
really_inline explicit operator simd8<uint8_t>() const { return vreinterpretq_u8_s8(this->value); }
```

with

```
really_inline explicit operator simd8<uint8_t>() const { return vreinterpretq_u8_s8(this); }
```

and I think it has to do with the fact that Visual Studio has `int8x16_t `and `uint8x16_t`.


It would be interesting to investigate whether we can get CI tests to run under ARM64 for appveyor but I did not look too deeply into it.

File: src/arm64/simd.h [hunk 1]
```diff
@@ -5,11 +5,94 @@
...
+
+
+/**
+ * make_uint8x16_t initializes a SIMD register (uint8x16_t).
+ * This is needed because, incredibly, the syntax uint8x16_t x = {1,2,3...}
+ * is not recognized under Visual Studio! This is a workaround.
+ * Using a std::initializer_list<uint8_t>  as a parameter resulted in
+ * inefficient code. With the current approach, if the parameters are
+ * compile-time constants,
+ * GNU GCC compiles it to ldr, the same as uint8x16_t x = {1,2,3...}.
```
Comment (jkeiser) [inline src/arm64/simd.h:27]: This is incredible. OMG.

File: src/arm64/simd.h [hunk 2]
```diff
@@ -61,8 +144,13 @@ namespace simd {
     // We return uint32_t instead of uint16_t because that seems to be more efficient for most
     // purposes (cutting it down to uint16_t costs performance in some compilers).
     really_inline uint32_t to_bitmask() const {
-      const uint8x16_t bit_mask = {0x01, 0x02, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80,
-                                   0x01, 0x02, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80};
+#ifdef SIMDJSON_REGULAR_VISUAL_STUDIO
+      const uint8x16_t bit_mask =  make_uint8x16_t(0x01, 0x02, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80,
+                                                   0x01, 0x02, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80);
+#else
+      const uint8x16_t bit_mask =  {0x01, 0x02, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80,
```
Comment (jkeiser) [inline src/arm64/simd.h:151]: Not a change request at all, but does make_uint8x16_t() generate optimal code on other platforms too, so we wouldn't have to ifdef?

File: src/arm64/simd.h (hunk 1 above)
Comment (lemire) [reply src/arm64/simd.h:27]: @jkeiser  I'd be happy to be proven wrong. Pull out Visual Studio, install the ARM thingies and try it.

File: src/arm64/simd.h (hunk 2 above)
Comment (lemire) [reply src/arm64/simd.h:151]: > Not a change request at all, but does make_uint8x16_t() generate optimal code on other platforms too, so we wouldn't have to ifdef?

I think it does but I am not sure. I was getting performance warnings so I got nervous and decided to go full on paranoid.

**Comment by lemire** [comment]
Ok. I will merge this and we can improve upon it. I will be very happy if someone manages to make it cleaner, because this is not pretty.

**Comment by lemire** [comment]
@jkeiser You'll be sad that it makes appveyor even slower. But I think it is for a good cause.

**Pull request merged**